torch_geometric/distributed/rpc.py

```diff
@@ -0,0 +1,212 @@
+
+            worker_list[node_rank] = worker_name
+            _rpc_worker_names[role] = worker_list
+
+        global_barrier(timeout=rpc_timeout)
+
+
+def shutdown_rpc(graceful=True):
+    if rpc_is_initialized() is True:
+        rpc.shutdown(graceful=graceful)
+
+
+atexit.register(shutdown_rpc, False)
+
+## RPC synchronization and routing with data partition mapping.
+
+
+class RpcRouter(object):
+    r""" A router to get the worker based on partition id."""
+    def __init__(self, partition2workers: List[List[str]]):
+        for pidx, rpc_worker_list in enumerate(partition2workers):
+            if len(rpc_worker_list) == 0:
+                raise ValueError("no rpc worker is in worklist'.")
+        self.partition2workers = partition2workers
+        self.rpc_worker_indexs = [0 for _ in range(len(partition2workers))]
+
+    def get_to_worker(self, data_partition_idx: int) -> str:
+        rpc_worker_list = self.partition2workers[data_partition_idx]
+        worker_idx = self.rpc_worker_indexs[data_partition_idx]
+        router_worker = rpc_worker_list[worker_idx]
+        self.rpc_worker_indexs[data_partition_idx] = \
+            (worker_idx + 1) % len(rpc_worker_list)
+        return router_worker
+
+
+@_require_initialized
+def rpc_partition2workers(num_data_partitions: int,
+                          current_partition_idx: int):
+    r""" all_gather to get the mapping between partition and workers """
+    ctx = get_context()
+    partition2workers = [[] for _ in range(num_data_partitions)]
+    gathered_results = global_all_gather(
+        (ctx.role, num_data_partitions, current_partition_idx))
+    for worker_name, (role, nparts, idx) in gathered_results.items():
+        partition2workers[idx].append(worker_name)
+    return partition2workers
+
+
+class RpcCallBase(ABC):
+    r""" A wrapper base for rpc call in remote processes."""
+    def __init__(self):
+        pass
+
+    @abstractmethod
+    def rpc_sync_call(self, *args, **kwargs):
+        pass
+
+    @abstractmethod
+    def rpc_async_call(self, *args, **kwargs):
+        pass
+
+
+_rpc_call_lock = threading.RLock()
+_rpc_call_id: int = 0
+_rpc_call_pool: Dict[int, RpcCallBase] = {}
+
+
+@_require_initialized
+def rpc_register(call: RpcCallBase):
+    r""" Register a call for rpc requests """
+    global _rpc_call_id, _rpc_call_pool
+    with _rpc_call_lock:
+        call_id = _rpc_call_id
+        _rpc_call_id += 1
+        if call_id in _rpc_call_pool:
+            raise RuntimeError(f"'rpc_register': register with the twice.")
+        _rpc_call_pool[call_id] = call
+
+    return call_id
+
+
+def _rpc_remote_async_call(call_id, *args, **kwargs):
+    r""" Entry for rpc requests """
+    return _rpc_call_pool.get(call_id).rpc_async_call(*args, **kwargs)
+
+
+def _rpc_remote_sync_all(call_id, *args, **kwargs):
+    r""" Entry for rpc requests """
+    return _rpc_call_pool.get(call_id).rpc_sync_call(*args, **kwargs)
```

Comment:
Why do we need a global pool here? Can't we just pas the sync or async function and its args directly to rpc_request_async and rpc_request_sync below?